safecast_deploy/state.py

```diff
@@ -1,104 +1,120 @@
 import boto3
+import collections
+import dataclasses
 import re
 import sys
 
+from safecast_deploy import aws_state
+from safecast_deploy.aws_state import AwsState, AwsTier, ParsedVersion, EnvType, AwsTierType
+from safecast_deploy.exceptions import InvalidVersionException
+from functools import lru_cache
 
-class State:
-    def __init__(
-            self,
-            app,
-            env=None,
-            new_version=None,
-            new_arn=None):
-        self.app = app
-        self.env = env
-        self.new_version = new_version
-        self.new_arn = new_arn
 
-        self.subenvs = {
-            'web': env,
-            'wrk': '{}-wrk'.format(env),
-        }
+class State:
+    def __init__(self, aws_app_name, eb_client):
+        self.aws_app_name = aws_app_name
+        self._c = eb_client
 
-        self.eb_client = boto3.client('elasticbeanstalk')
-        self._c = self.eb_client
-        self._identify_current_envs()
         self._classify_available_versions()
-        self._validate_version()
 
-    def _validate_version(self):
-        if self.new_version is None:
+        self.old_aws_state = AwsState(
+            aws_app_name=self.aws_app_name,
+            envs=self._build_envs(),
+        )
+
+    @lru_cache(typed=True)
+    def new_aws_state(self, target_env_type, new_version=None, new_arn=None):
+        self._validate_version(new_version)
+        parsed_version = self._parse_version(new_version)
+        new_envs = collections.defaultdict(dict)
+        new_env_num = max([(tier.num + 1) % 1000 for (tier_type, tier) in self.old_aws_state.envs[target_env_type]], key=operator.itemgetter(1))
+        for (tier_type, tier) in self.old_aws_state.envs[target_env_type]:
```

Comment:
```suggestion
        new_env_num = max([(tier.num + 1) % 1000 for (tier_type, tier) in self.old_aws_state.envs[target_env_type].items()])
        for (tier_type, tier) in self.old_aws_state.envs[target_env_type].items():
```

This got me a bit further. But then I hit this

```
  File "/Users/matschaffer/safecast/api/safecast_deploy/deploy.py", line 183, in run_same_env
    state.new_aws_state(env_type, new_version=args.version),
  File "/Users/matschaffer/safecast/api/safecast_deploy/safecast_deploy/state.py", line 36, in new_aws_state
    new_tier = dataclasses.replace(new_tier, version=new_version, parsed_version=parsed_version)
  File "/Users/matschaffer/.asdf/installs/python/3.8.5/lib/python3.8/dataclasses.py", line 1293, in replace
```

Looks like AwsTier doesn't have a `version`. Not sure what the intent is

Reply:
github might be confused on the diff. Was trying to suggest for lines 30/31

Reply:
It's probably going to be easier for me to just write the unit test for that method and shake out all the issues.

And then move to a static-typed, compiled language. Although I hear Python has some static type support in 3.6 that can be used for checking before execution.

Reply:
I've been learning rust and typescript lately. Happy to try either :)

Reply:
Now that I've got more experience with it, I think it would be possible to build a Terraform provider to do much of this and automatically determine whether to use a new environment (although the official AWS ones would not be sufficient, would have to rewrite from scratch). Not that I'm planning another rewrite.